docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,256 @@
+Bottlerocket will iteratively invoke generators until all generation has completed, or in cases where it seems likely that a deadlock has occurred.
+
+```toml
+# setting.toml requires the network and kubernetes settings before it can emit its own
+[generation.required]
+network = v1
+kubernetes = v2
+```
+
+## Settings Extension Development and the Settings SDK
+Settings extensions can be written in any programming language, so long as a symlink to a binary speaking the "extension protocol" lands in the right spot on the filesystem.
+That said, many settings extensions need to tackle the same challenges: How do we represent and implement settings versions? Migrations? Validation? Template utilities? Rather than forcing all extensions to implement this in isolation, we will create a shared library, called the "Settings SDK", which will be built in Bottlerocket's monorepo and published to crates.io.
+
+The goal of the settings SDK is to provide standard functions and macros for interfacing with the settings extension protocol.
+While our model here provides quite a lot of power to settings extension developers, the SDK would allow us to exert some pressure in the form of opinions on simple ways to manage data or perform migrations.
+
+As an example, the SDK will provide a Rust trait which can be implemented for each model in order to adequately implement all settings extensions methods on a command line interface.
+Once those traits are implemented, you might implement this CLI as such:
+
+```rust
+fn main() -> Result<()> {
+    bottlerocket_settings_sdk::SettingsExtension::with_models(vec![
+        BottlerocketSetting::<v1::MotdV1>::model(),
+        BottlerocketSetting::<v2::MotdV2>::model(),
+    ])
+    .run_extension()
+    .context("Settings extension encountered an error.")
+}
+```
+
+## Asked and Anticipated Questions
+### Why not use some existing schema or modeling language to define the shape of settings?
+Existing modeling languages are powerful, but even in our existing Bottlerocket settings we have validations which we have implemented as functions in the Rust code (or otherwise wished we had a simple way to do so.) Rather than strip power from settings extensions by dictating that they use one of these modeling languages, we leave it entirely up to the binary to decide how best to validate inputs.
+In many cases, this may mean that the binaries carry with them their own implementation of a popular schema language, like JSONschema.
+
+### How will the `apiserver` handle concurrent requests in the face of transactional writes?
+The `apiserver` will use a [filesystem-based](https://man7.org/linux/man-pages/man2/flock.2.html) reader-writer lock to ensure that the datastore is only accessed by concurrent readers or a single writer. Writes which are sent to the `apiserver` while the file is lock are placed in a queue.
+
+## Appendices
+### A. Assisted EC2 Userdata Generation and Validation
```

Comment:
Would it be better to tackle this in a follow-up design? I agree that it's a potential benefit, but it seems orthogonal to the main concepts here.

Reply:
Totally fair. It's included here to capture some verbal discussion of this document that occurred and I wanted to capture it as a possible feature of the new system, but it can be delegated to another doc.